src/vec.rs

```diff
@@ -0,0 +1,364 @@
+/// +--------------------------------------+
+///
+/// T::fixed_value_size = false
+/// +------------------------------------------+
+/// | header | [size; val] | [size; val] | ... |
+/// +------------------------------------------+
+/// size: u32
+///
+/// Returns the pointer to the entry (read: not necessarily the value).
+pub struct Vec<M, T> {
+    memory: M,
+    max_value_size: u64,
+    base: Address,
+
+    len: usize,
+
+    // A marker to communicate to the Rust compiler that we own these types.
+    _phantom: PhantomData<T>,
+}
+
+#[repr(packed)]
+struct StableVecHeader {
+    magic: [u8; 3],
+    version: u8,
+    max_value_size: u64,
+    fixed_value_size: bool,
+    len: usize,
+    // Additional space reserved to add new fields without breaking backward-compatibility.
+    _buffer: [u8; 24],
+}
+
+impl StableVecHeader {
+    fn size() -> Bytes {
+        Bytes::from(core::mem::size_of::<Self>() as u64)
+    }
+}
+
+impl<M: Memory, T: BoundedStorable> Vec<M, T> {
+    const ENTRY_SIZE: usize = (T::MAX_SIZE + if T::FIXED_SIZE { 0 } else { 4 }) as usize;
+
+    #[must_use]
+    pub fn new(memory: M) -> Self {
+        Self::new_with_sizes(memory, T::MAX_SIZE as u64)
+    }
+
+    #[must_use]
+    pub fn new_with_sizes(memory: M, max_value_size: u64) -> Self {
+        let v = Self {
+            memory,
+            max_value_size,
+            base: Address::from(0) + StableVecHeader::size(),
+            len: 0,
+            _phantom: PhantomData,
+        };
+
+        v.save();
+        v
+    }
+
+    pub fn load(memory: M) -> Self {
+        Self::load_with_sizes(memory, T::MAX_SIZE as u64)
+    }
+
+    pub fn load_with_sizes(memory: M, max_value_size: u64) -> Self {
+        // Read the header from memory.
+        let header: StableVecHeader = read_struct(Address::from(0), &memory);
+        assert_eq!(&header.magic, MAGIC, "Bad magic.");
+        assert_eq!(header.version, LAYOUT_VERSION, "Unsupported version.");
+        assert_eq!(header.fixed_value_size, T::FIXED_SIZE);
+        let expected_value_size = header.max_value_size;
+        assert!(
+            max_value_size <= expected_value_size,
+            "max_value_size must be <= {}",
+            expected_value_size
+        );
+
+        Self {
+            memory,
+            max_value_size: header.max_value_size,
+            base: Address::from(0) + StableVecHeader::size(),
+            len: header.len,
+            _phantom: PhantomData,
+        }
+    }
+
+    fn save(&self) {
+        let header = StableVecHeader {
+            magic: *MAGIC,
+            version: LAYOUT_VERSION,
+            max_value_size: self.max_value_size,
+            fixed_value_size: T::FIXED_SIZE,
+            len: self.len,
+            _buffer: [0; 24],
+        };
+
+        write_struct(&header, Address::from(0), &self.memory);
+    }
+
+    pub fn clear(&mut self) {
+        self.len = 0;
+        self.save();
+    }
+
+    #[inline]
```

Comment:
I'm not sure I understand why we're inlining some of these methods. This is a private method, so I'd expect the compiler to know whether or not to inline it. Is this based on some benchmarking?